File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

/*
전력망을 2개로 나눴을 때 가장 둘이 비슷하게 된 케이스 조사
저기 wires에서 하나씩만 빼고 계산 (8개 원소)을 하고
그룹 몇개로 나오는지 조사
*/

int mini=1e9;

bool isVisited[1000];

void bfs(int n, vector<vector<int>> line){
    
    int cnt=1;
    
    queue<int> q;
    
    q.push(1);
    isVisited[1]=true;
    
    while(!q.empty()){
        
        
        int cur=q.front();
        q.pop();
        
        for(int k=0; k<line[cur].size(); k++){
            
            int value=line[cur][k];
            
            if(isVisited[value]) continue;
            
            isVisited[value]=true;
            
            q.push(value);
            
            cnt++;
        }
        
    }
    
    mini= min(mini, abs((n-cnt)-cnt));
}

void cal(int n, vector<vector<int>> wires) {
    
    for(int i=0; i<wires.size(); i++){
        
        vector<vector<int>> line(n+1);
        
        for(int j=0; j<wires.size(); j++){
            
            if(i==j) continue;
            
            int fir=wires[j][0];
            int sec=wires[j][1];
            
            line[fir].push_back(sec);
            line[sec].push_back(fir);
        }
        
        for(int i=0; i<n+1; i++) isVisited[i]=false;
        
        bfs(n, line);
    }
    
}

int solution(int n, vector<vector<int>> wires) {
    
    cal(n, wires);

    return mini;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
import java.io.*;

class Solution {
    
    int answer=Integer.MAX_VALUE;
     
    void bfs(int n, int[][] wires, int exclude){
        
        boolean[] isVisited = new boolean[n+1];
        
        List<List<Integer>> lists = new ArrayList<>();
        
        for(int i=0; i<=n; i++) lists.add(new ArrayList<>());
        
        for(int i=0; i<wires.length; i++){
            if(i==exclude) continue;
            int first=wires[i][0];
            int second=wires[i][1];
            
            lists.get(first).add(second);
            lists.get(second).add(first);
        }
        
        Queue<Integer> q = new LinkedList<>();
        
        int cnt=1;            
        q.offer(1);
        isVisited[1]=true;
        
        while(!q.isEmpty()){
            int curQ=q.poll();
            
            for(int edge : lists.get(curQ)){
                if(isVisited[edge]) continue;
                
                isVisited[edge]=true;
                cnt++;
                q.offer(edge);
            }
        }
        int absValue=Math.abs(n-2*cnt);
        answer=Math.min(answer,absValue); 
    }
    
    public int solution(int n, int[][] wires) {
        
        for(int i=0; i<n; i++) bfs(n,wires,i);
        
        return answer;
    }
}